Improve selftest of JsonPreprocessor package

namsonx opened the pull request:
Update selftest using pytest for JsonPreprocessor package

namsonx committed:
Improve selftest of JsonPreprocessor package

HolQue commented:
Hi Son,

I wanted the python-jsonpreprocessor repository to support the TestTrigger mechanism. Therefore I added a TestExecutor 'executerobottest.py' to the atest folder.

https://github.com/test-fullautomation/python-jsonpreprocessor/pull/31

The execution causes the following error message:
File 'D:/RobotFramework/python39/lib/site-packages/testdata/config/07_json_format/json_format_01.json' is not existing!
It's interesting that the testdata are expected to be present within the site-packages.

This happens in all tests. Therefore all tests fail.

The atest folder already contains a script 'run.py' that also executes the test. With 'run.py' this problem does not occur. Most of the tests are passed, except "Test Sub Data Structure 06":

Test Sub Data Structure 06 :: Updated more than 1 parameter with n... | FAIL |
{'preproStructure': {'addNewPreParam': 'Adds new param', 'newStruct': {'check1': 1, 'check2': 2}, 'variable_01': '1'}, 'preproTest': {'param_01': 'Updated new value'}} != {'preproStructure': {'addNewPreParam': 'Adds new param', 'newStruct': {'check1': 1, 'check2': 2}, 'variable_01': 1}, 'preproTest': {'param_01': 'Updated new value'}}

The main difference between executerobottest.py and run.py is: run.py uses robot.run() to execute the tests, whereas executerobottest.py uses subprocess.call().

And executerobottest.py does some things more, like the handling the algebraic sign of the return value and passing command lines to the Robot Framework.

We have now two possibilities:

1. Extend run.py by the missing features
2. Fix the import path problem of JSON files to enable the execution of executerobottest.py

I would prefer the second option because this would make it possible to handle the test execution in every repository in the same way with the same mechanisms, like described in

https://github.com/test-fullautomation/RobotFramework_AIO/blob/develop/test/documentation/RF(AIO)-TestConcept.pptx

It would also be interesting to learn about the root cause for this deviating behavior.

HolQue committed:
Added executerobottest.py to support the TestTrigger

namsonx commented:
Hello Holger,

I already deleted 'run.py' and replaced by 'run.bat' for Windows and 'run.sh' for linux.
Currently, we just execute 'run.bat' or 'run.sh' to run selftest, and selftest now develop using python instead of robot.
Do we have any test mechanism for pytest?

HolQue committed:
Replaced executerobottest.py by executepytest.py (because of planned rework of the atest)

HolQue commented:
Hi Son,

for executing pytests you have to use the executepytest.py

I have updated 
https://github.com/test-fullautomation/python-jsonpreprocessor/pull/31
to make this script available for you.

test-fullautomation commented:
Hi both,
is your problem test execution problem solved? 
Thank you,
 Thomas

test-fullautomation commented on atest/testdata/config/06_override_params/project_config_02.json:
```diff
@@ -29,7 +25,7 @@
    "${gPreproString}": "Override string param in imported file.",
    "params":{
-     "global":{
+     "glob":{
```
Hi  Son,
where is "glob" coming from?
I was not asked for this and I did therefore not agree to this change.
I'm sorry that you need to do some rework now.

Python has a "global" keyword for global parameters:
https://www.w3schools.com/python/python_variables_global.asp

We use what users expect. Therefore "global" is the correct key here to indicate a global scope.

In addition "glob" has another meaning in english language.

Thank you,
 Thomas

namsonx replied:
Hello Thomas,

Due to the "global" is reserved keyword of Python, so some scenarios of pytest failed when I name the key of Json object is "global".
That the reason I just changed this key name of Json configuration file.

Thank you,
Son

test-fullautomation replied:
Hi Son,
O.K. I understand your motivation. But this is no solution to make tests running.
Can not be that the test tool makes a global syntax change required.,,
Can you please discuss with Cuong if he has an idea what to do.
Thank you,
Thomas

test-fullautomation replied:
Hi Son,
what is the status here?
Thank you,
 Thomas

namsonx replied:
Hello Thomas,

I created [new commit ](https://github.com/test-fullautomation/python-jsonpreprocessor/pull/30/commits/5a7d154f3e29f39a894ef6682f2619a6801f00cc) to use object name 'global' instead of 'glob'.
I also run this selftest on both Windows and OSD6.

Thank you,
Son

test-fullautomation commented:
Hi Son,
I like the structure and way you do the testing now. This is a very good proposal.
Please look into my findings.
Thank you,
 Thomas

test-fullautomation commented on atest/testdata/config/06_override_params/project_config_04.json:
```diff
@@ -0,0 +1,67 @@
+      "minorversion": "1",
+      "patchversion": "0"
+    },
+    "Maximum_version": "0.5.3",
```
Hi Son,
Maximum Version must be minimum 0.6.0
Thank you, 
Thomas

namsonx replied:
Hi Thomas,

I will update new Maximum Version.

Thank you,
Son

test-fullautomation commented on atest/testdata/templates.py:
```diff
@@ -501,4 +501,33 @@
+      }
+    }
+  },
+  "preprocessor": {
```
Hi Son,
preprocessor data and tests has to be removed as long as we don't provide a corresponding feature.
Thank you,
Thomas

namsonx replied:
Hi Thomas,

This is just test data for JsonPreprocessor package selftest and as we define JsonPreprocessor and RobotFramework TestsuitesManagement are 2 independent packages. This test data is not related to the feature which providing in TestsuitesManagement, so I think we can keep this test data as it is for JsonPreprocessor selftest.

Thank you,
Son

test-fullautomation replied:
Hi Son,
accepted!
Thank you,
 Thomas

namsonx committed:
Updated testdata of JsonPreprocessor selftest

test-fullautomation reviewed (changes requested):
Hi Son,

please check my findings.

Thank you,
 Thomas

HolQue committed:
tiny fixes

namsonx committed:
Update documentation base on new version

namsonx committed:
Update to reduce redundant parts of documentation

ngoan1608 committed:
fix document generation issue due to redundant leading space in section header

test-fullautomation commented:
Hi Son,
what is the status here?
Thank you,
 Thomas

namsonx committed:
Updates selftest data according to review comment

test-fullautomation reviewed (dismissed):
Hi Son,

looks good to me except of the use of "glob" instead of "global".

What is the status here?

Thank you,
 Thomas

namsonx commented:
> Hi Son,
> 
> looks good to me except of the use of "glob" instead of "global".
> 
> What is the status here?
> 
> Thank you, Thomas

Hello Thomas,

I created [new commit ](https://github.com/test-fullautomation/python-jsonpreprocessor/pull/30/commits/5a7d154f3e29f39a894ef6682f2619a6801f00cc) to use object name 'global' instead of 'glob'.
I also run this selftest on both Windows and OSD6.

Thank you,
Son

namsonx committed:
Fix issue40 - single quotes error

namsonx committed:
Fix issue34 - absolute paths in sample

namsonx committed:
Adds encoding when load config files to support utf8 characters

test-fullautomation commented:
Hi Son,
very nice!

How did you solve the "glob" -> "global" problem now. I didn't find the place in the changed code.

Thank you,
 Thomas

test-fullautomation reviewed (approved):
Hi Son,
looks good to me now!
Thank you,
 Thomas
